Add authenticated api requests support to query multiple endpoints

dbollaer opened the pull request:
Thank you for providing this authentication source.
If possible, I would like to contribute.

I need this new functionality to query multiple endpoints in azureAD.
We need the personal info and the memberOf.

I did not understand how test this new piece of code, and would welcome your feedback.

dbollaer commented:
Hello @pradtke, could you please have look? Thank you in advance.

pradtke commented on lib/Auth/Source/OAuth2.php:
```diff
@@ -213,6 +221,22 @@ function () use ($provider, $accessToken) {
+            },
+                $this->config->getInteger('retryOnError', 1)
+            );
+            $attributes = array_replace_recursive($attributes, $apiAttributes);
```
The array structure of apiRequest may not be compatible with SSP attributes (e.g. it may have values that are maps). You need to call `convertResourceOwnerAttributes` to flatten the data structure.

dbollaer replied:
I think this is doen line 241.

pradtke replied:
oh yeah, you're correct.

pradtke commented:
For testing, first confirm that running `./vendor/bin/phpunit` works without an issue. Then add a new test case to `OAuth2Test` that is similar to the method `testFinalSteps`. You'll need to mock your interaction with the provider `getAuthenticatedRequest`. There is an example of how to do that in `LinkedInV2AuthTest`  class,  test case `testGettingEmail`.  I think you'd want to test that a) both simple and complex response arrays work, b) that attribute merging works the way you expect if the same attribute appears more than once.

Thanks for the contribution!

dbollaer commented:
@pradtke I added additional tests and cleaned up my code a little.
I also changed the logic for postFinalStep. Now if the attributes already exists, they have preference.
The logic is that the graph API calls have priority.
The reason for this change: azureAD, sends only the first 150 security groups in the token.
I need to be able to retrieve all security groups.
Please advice, on next steps.
Thank you in advance.

pradtke commented on lib/Auth/Source/OAuth2.php:
```diff
@@ -213,6 +204,45 @@ function () use ($provider, $accessToken) {
 
         $attributes = $resourceOwner->toArray();
+
+        if (
```
why did this block move here, and have the `logIdTokenJson` condition removed?

dbollaer replied:
Was for easier debugging, I put it back. Sorry for inconvenience.

pradtke commented on lib/Auth/Source/MicrosoftHybridAuth.php:
```diff
@@ -48,10 +48,10 @@ protected function postFinalStep(AccessToken $accessToken, AbstractProvider $pro
             $state['Attributes'][$prefix . 'mail'] = [$idTokenData['email']];
         }
-        if (array_key_exists('name', $idTokenData)) {
+        if (array_key_exists('name', $idTokenData) && !array_key_exists($prefix . 'name' , $state['Attributes'])) {
```
I didn't understand the motivation for this change. I'd prefer not to make it - we validated the existing code with a mix of live, converged and azure ad apps  x  social, guest and azure ad accounts and I don't want to redo that validation :)  I recall that MS was _weird_ in how attributes were returned in some of those situation

dbollaer replied:
I returned it to its original state.

pradtke commented:
@dbollaer I had two last comments.  I think everything else looked pretty good.

dbollaer commented:
@pradtke 
Thank you for your feedback.
I am using a docker container to run my php tests. Because I am having some trouble with running php locally. Would it be possible to add it to the documentation.
I think it would make it easier for other people to contribute.

```
docker run --rm -it  -v $PWD:$PWD -w=$PWD alterway/php:7.2-cli /usr/local/bin/composer install
docker run --rm -it  -v $PWD:$PWD -w=$PWD alterway/php:7.2-cli ./vendor/bin/phpunit
```

Thank you again for all help.

dbollaer commented:
@pradtke small question, when do you think the next release will be?
Thank you.

pradtke commented:
@dbollaer I released a new version 3.1 with this change
